src/data_gradients/datasets/base_dataset.py

```diff
@@ -0,0 +1,88 @@
+import os
+from abc import ABC, abstractmethod
+from typing import Tuple, Sequence, Optional
+
+import numpy as np
+from torch.utils.data.dataset import Dataset
+from data_gradients.datasets.FolderProcessor import ImageLabelFilesIterator, DEFAULT_IMG_EXTENSIONS
+from data_gradients.datasets.utils import load_image, ImageChannelFormat
+
+
+class BaseDataset(Dataset, ABC):
```

Comment:
Let's imagine we are adding NLP task at some point. Would that base class be suitable for this task with `load_image` and `load_labels` methods that returns numpy arrays?

Second complain I have is that API is not clear at all.
* Why there is `__iter__` and `__getitem__`? Which one should I use (prefer)?
* Why there are separate methods `load_image` and `load_labels` and why one have to implement them explicitly if he just implemented `__getitem__`?

Third problem is that we don't use ImageSample / SegmentationSample / DetectionSample and instead use "plain" numpy arrays without any metadata and clear structure what is stored inside such arrays.

Reply:
Well, yeah I guess this abstract class is not that useful anyway, I think we can just take it out.

About `load_image` and `load_labels` there is a difference, which is that `load_image` can be easily be implemented for one class and be shared among any child (which is what I did), which means that we only need to implement `load_labels` afterward. Besides, having it separated to `__getitem__` was motivated to avoid mixing responsabilities;  for instance if we want to add a transform step, this should not be part of `load_labels`, but should be done inside `__getitems__` instead. But I agree that this should not be part of the most basic abstract class.

Concerning the samples, whether we should return these objects or not is a good question.
My approach to this was "Let's implement simple datasets similar to what users may use (return image/label) so that users can also play around with it easily", also motivated by the simplicity and low coupling this would create with the rest of the code. But I also see clear advantages to what you proposed:

Pro:
- More explicit output
- (In theory) no need to go through the `batch_processor` steps. In practice, we would need to update the design for that, because currently any dataset output is passed to the `batch_processor`.
- Abstracting the output of the dataset into these `Sample` classes makes these datasets easy to use in different context without needing to worry how the dataset outputs data.

Con:
- The current pipeline expects a dataset/dataloader/generator that yields samples in dict/list/tuple/... format, not dataclass. To resolve this, I think the clean way would be to check if the user dataset is a DG dataset; If yes, take the dataset output as is, if not we wrap the user dataset with the `batch_processor`. Introducing a clean "wrapping" class  would mean more work because the current pipelines are a bit coupled with the `batch_processor`.
- [Most importantly] Not compatible with dataloaders (dict/tuple/list can be collated, not dataclasses). We can still define custom collate functions, but it is definitely less nice for users

Any other thing that you think should be considered? And don't you think that the dataloader compatibility is a bit problematic for a `Dataset` ?

Reply:
Ok, I now have even more questions that before..

> About load_image and load_labels there is a difference, which is that load_image can be easily be implemented for one class and be shared among any child (which is what I did), which means that we only need to implement load_labels afterward. 

Composition is better than inheritance.

> (In theory) no need to go through the batch_processor steps. In practice, we would need to update the design for that, because currently any dataset output is passed to the batch_processor.

Like, yes, updating design is a right thing if design is broken


> And don't you think that the dataloader compatibility is a bit problematic for a Dataset ?

But why we have to use dataloader at all in the first place here? This is somewhat making task unnecessary more complicated to me. If we already have the dataset class - just iterate over it and that's it. 

For who this new class is for? Is it for use, for internal API or we're going to ask user to inherit from this class? If feel that this is a part of internal API, but at the same time I'm not sure about it.